### DISCUSSION TITLE
CPU usage Plugdata vs Pure Data

### DISCUSSION THREAD

**Discussion by sevdh** [plugdata-team/plugdata#616] [General]
It seems to me that Plugdata uses quite a bit more CPU than Pure Data for the same patch. To do a first test I've made a patch that contain many bob~ objects and without updating GUI objects.

I'm using a macbook pro M1 running macos 12.6.3.

- When running 100 bob~'s the audio in Plugdata gets silent(!) while the level meters at the bottom of the window do indicate there is audio! The energy usage in the activity monitor is less than 100% but this low number seems to be the result of Plugdata not being able to handle the load (and dropping things) resulting in silence.
- When running 60 bob~'s Plugdata can just deal with it. The energy usage in the activity monitor is now around 100%.
- Pure Data on the other hand can easily deal with 180 bob~ objects.
- What is nice in Plugdata is that you can select all objects and move them around in the window without any glitches but when you release the mouse button an unfortunate glitch in the audio occurs which I suppose is not needed.
- Pure Data, on the other hand, starts producing lots of glitches when you start moving the selected objects around the window. I guess this is because the interface runs in the same thread as the DSP.

Attached are the patches.

[bobs.zip](https://github.com/plugdata-team/plugdata/files/10863647/bobs.zip)

**Comment by timothyschoen** [comment]
> It seems to me that Plugdata uses quite a bit more CPU than Pure Data for the same patch. To do a first test I've made a patch that contain many bob~ objects and without updating GUI objects.
> 
> I'm using a macbook pro M1 running macos 12.6.3.
> 
>     * When running 100 bob~'s the audio in Plugdata gets silent(!) while the level meters at the bottom of the window do indicate there is audio! The energy usage in the activity monitor is less than 100% but this low number seems to be the result of Plugdata not being able to handle the load (and dropping things) resulting in silence.
> 
>     * When running 60 bob~'s Plugdata can just deal with it. The energy usage in the activity monitor is now around 100%.
> 
>     * Pure Data on the other hand can easily deal with 180 bob~ objects.
> 
>     * What is nice in Plugdata is that you can select all objects and move them around in the window without any glitches but when you release the mouse button an unfortunate glitch in the audio occurs which I suppose is not needed.
> 
>     * Pure Data, on the other hand, starts producing lots of glitches when you start moving the selected objects around the window. I guess this is because the interface runs in the same thread as the DSP.
> 
> 
> Attached are the patches.
> 
> [bobs.zip](https://github.com/plugdata-team/plugdata/files/10863647/bobs.zip)

I'm also using an M1 Macbook, so these patches are useful for me! 100 bobs still runs okay in plugdata, but 180 definitely doesn't, even though it's only at 80% CPU usage. I've also verified that there's no locking of the audio thread happening, that can't be it. I suspect this may be to do with the way JUCE/Pd handles audio IO, but I'm not sure.

The good thing is that I see that with many bobs, pretty much all the processor load is still coming from pd, and not much is coming from plugdata itself. So that also confirms that it has something to do with the way it handles audio.

I think the glitches in Pd are indeed because of Pd running the GUI and audio from the same thread. Pd sends gui commands over to a tcl/tk window over sockets, so if Pd gets slowed down because of audio tasks, it won't receive messages as frequently and things may get glitchy.

**Reply by FlachyJoe** [reply]
> I think the glitches in Pd are indeed because of Pd running the GUI and audio from the same thread. Pd sends gui commands over to a tcl/tk window over sockets, so if Pd gets slowed down because of audio tasks, it won't receive messages as frequently and things may get glitchy.

even launch without gui, pd glitchs
`pd -nogui bob-180.pd`

**Reply by sevdh** [reply]
I get up to 180 bobs with pd on a macbook pro m1. Other machines may of course have other limits. Since the patch makes minimal use of GUI elements I don't expect a real difference between gui and -nogui modes.

**Reply by sevdh** [reply]
The main reason for my post is that I only get up to 60 bobs in plugdata.

**Comment by sevdh** [comment]
Thanks for the reply!

When I overload Plugdata with too many bobs the CPU usage goes down and the audio becomes silent. Does that happen for you as well?

Peculiar that you can run 100 and I can run only 60 bobs in Plugdata! This results in a factor 3 difference for me between Plugdata and Puredata.

I'm testing with the very latest built.

**Reply by timothyschoen** [reply]
You're not building manually right? Because if that were the case, it would be important to make sure you're building in release mode. If you're grabbing the nightly builds from Github or the website, that can't be the problem.

> When I overload Plugdata with too many bobs the CPU usage goes down and the audio becomes silent. Does that happen for you as well?

Same for me, I can think of two possible reasons that I can investigate. I almost seems like JUCE just "gives up" when your audio routine is too intensive, so maybe that's what's going on.

One other explanation is that there might be some memory allocation on the audio thread slowing things down. That would mean that plugdata needs to wait for the OS to give it more memory, which would explain why there's lower CPU usage (it's just waiting), while also being glitchy.

**Reply by sevdh** [reply]
I'm downloading the builds from GitHub.
